rust/lance/src/index/vector/ivf/builder.rs

```diff
@@ -229,17 +163,40 @@ pub(super) async fn build_hnsw_sq_partitions(
     )
     .await?;
 
-    write_hnsw_sq_index_partitions(
+    write_hnsw_quantization_index_partitions(
         dataset,
         column,
         metric_type,
         hnsw_params,
         writer,
         auxiliary_writer,
         ivf,
-        sq,
+        quantizer,
         Some(stream),
         None,
     )
     .await
 }
+
+pub fn build_hnsw_model(
+    hnsw_params: HnswBuildParams,
+    vector_array: Vec<Arc<dyn Array>>,
+) -> Result<(HNSW, Arc<dyn Array>)> {
+    let vector_arrs = vector_array
+        .iter()
+        .map(|arr| arr.as_ref())
+        .collect::<Vec<_>>();
+    let fsl = arrow_select::concat::concat(&vector_arrs)?;
+    std::mem::drop(vector_array);
+
+    let mat = Arc::new(MatrixView::<Float32Type>::try_from(
+        fsl.as_fixed_size_list(),
+    )?);
+
+    // We have normalized the vectors if the metric type is cosine, so we can use the L2 distance
```

Comment:
what about Dot? I am fine to just return explicit Error to say that `Dot is not supported in HNSW yet`

Reply:
just added a check before trying to build the index, so we can get the error before shuffling the dataset